domain/storage/state/state.go

```diff
@@ -0,0 +1,332 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	domainstorage "github.com/juju/juju/domain/storage"
+	"github.com/juju/juju/internal/uuid"
+)
+
+// State represents database interactions dealing with block devices.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new block device state
+// based on the input database factory method.
+func NewState(factory coredatabase.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+type poolAttributes map[string]string
+
+// CreateStoragePool creates a storage pool, returning an error satisfying [errors.AlreadyExists]
+// if a pool with the same name already exists.
+func (st State) CreateStoragePool(ctx context.Context, pool domainstorage.StoragePoolDetails) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	selectUUIDStmt, err := sqlair.Prepare("SELECT &StoragePool.uuid FROM storage_pool WHERE name = $StoragePool.name", StoragePool{})
+	if err != nil {
+		return errors.Trace(domain.CoerceError(err))
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		dbPool := StoragePool{Name: pool.Name}
+		err := tx.Query(ctx, selectUUIDStmt, dbPool).Get(&dbPool)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Trace(domain.CoerceError(err))
+		}
+		if err == nil {
+			return fmt.Errorf("storage pool %q %w", pool.Name, errors.AlreadyExists)
+		}
+		poolUUID := uuid.MustNewUUID().String()
+
+		if err := upsertStoragePool(ctx, tx, poolUUID, pool.Name, pool.Provider); err != nil {
+			return errors.Annotate(domain.CoerceError(err), "creating storage pool")
+		}
+
+		if err := updatePoolAttributes(ctx, tx, poolUUID, pool.Attrs); err != nil {
+			return errors.Annotatef(domain.CoerceError(err), "creating storage pool %s attributes", poolUUID)
+		}
+		return nil
+	})
+
+	return errors.Trace(err)
```

Comment:
I think CoerceError here and get rid of all the ones above. Just my two cents though.